graph/unmarshal.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2024 The Serverless Workflow Specification Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"sigs.k8s.io/yaml"
+
+	"github.com/serverlessworkflow/sdk-go/v4/internal/util"
+)
+
+// TODO: Remove global variable
+var HttpClient = http.Client{Timeout: time.Duration(1) * time.Second}
+
+func unmarshalNode(n *Node, data []byte) error {
+	data = bytes.TrimSpace(data)
+	if data[0] == '{' {
+		return unmarshalObject(n, data)
+	} else if data[0] == '[' {
+		return unmarshalList(n, data)
+	}
+
+	return json.Unmarshal(data, &n.value)
+}
+
+func unmarshalObject(n *Node, data []byte) error {
+	dataMap := map[string]json.RawMessage{}
+	err := json.Unmarshal(data, &dataMap)
+	if err != nil {
+		return err
+	}
+
+	for key, val := range dataMap {
+		node := n.Edge(key)
+		err := json.Unmarshal(val, &node)
+		if err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+func unmarshalList(n *Node, data []byte) error {
+	dataMap := []json.RawMessage{}
+	err := json.Unmarshal(data, &dataMap)
+	if err != nil {
+		return err
+	}
+
+	n.List(true)
+
+	for i, val := range dataMap {
+		key := fmt.Sprintf("%d", i)
+		node := n.Edge(key)
+		err := json.Unmarshal(val, &node)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func loadExternalResource(url string) (b []byte, err error) {
+	index := strings.Index(url, "://")
+	if index == -1 {
+		b, err = getBytesFromFile(url)
+	} else {
+		scheme := url[:index]
+		switch scheme {
+		case "http", "https":
+			b, err = getBytesFromHttp(url)
+		case "file":
+			b, err = getBytesFromFile(url[index+3:])
+		default:
+			return nil, fmt.Errorf("unsupported scheme: %q", scheme)
+		}
+	}
+	if err != nil {
+		return
+	}
+
+	// TODO: optimize this
```

Comment:
Are we planning to fix the TODOs in this PR or in upcoming PRs? If so we should create issues to address them.